xarray/testing/strategies.py

```diff
@@ -0,0 +1,761 @@
+        Function which accepts one numpy array and returns one numpy-like array of the same shape.
+        Applied to the data after it is drawn from the `data` strategy provided.
+        Useful for converting numpy arrays to other types of arrays, e.g. sparse arrays.
+        Default is a no-op.
+    """
+
+    if any(
+        not isinstance(arg, st.SearchStrategy) and arg is not None
+        for arg in [data, dims, attrs]
+    ):
+        raise TypeError(
+            "Contents must be provided as a hypothesis.strategies.SearchStrategy object (or None)."
+            "To specify fixed contents, use hypothesis.strategies.just()."
+        )
+
+    if data is not None and dims is None:
+        # no dims -> generate dims to match data
+        data = draw(data)
+        dims = draw(dimension_names(min_ndims=data.ndim, max_ndims=data.ndim))
+
+    elif dims is not None and data is None:
+        # no data -> generate data to match dims
+        dims = draw(dims)
+        if isinstance(dims, List):
+            valid_shapes = npst.array_shapes(min_dims=len(dims), max_dims=len(dims))
+            data = draw(np_arrays(shape=draw(valid_shapes)))
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            shape = tuple(dims.values())
+            data = draw(np_arrays(shape=shape))
+
+    elif data is not None and dims is not None:
+        # both data and dims provided -> check drawn examples are compatible
+        dims = draw(dims)
+
+        # TODO is there another way to enforce these assumptions? This is very like to fail hypothesis' health checks
+        # TODO how do I write a test that checks that the hypothesis Unsatisfiable error will be raised?
+        # TODO or we could just raise in this case?
+        if isinstance(dims, List):
+            data = draw(data)
+            assume(data.ndim == len(dims))
+        else:
+            # should be a mapping of form {dim_names: lengths}
+            data = draw(data)
+            shape = tuple(dims.values())
+            assume(data.shape == shape)
+
+    else:
+        # nothing provided, so generate everything consistently by drawing dims to match data
+        data = draw(np_arrays())
+        dims = draw(dimension_names(min_ndims=data.ndim, max_ndims=data.ndim))
+
+    if isinstance(attrs, st.SearchStrategy):
+        attrs = draw(attrs)
+    elif attrs is None:
+        # TODO autogenerate some attributes
+        ...
+
+    return xr.Variable(dims=dims, data=convert(data), attrs=attrs)
+
+
+El = TypeVar("El")
+
+
+# Mostly from the unfinished PR https://github.com/HypothesisWorks/hypothesis/pull/1533
+# TODO Should move this function upstream by opening new PR
+@st.composite
+def subsequences_of(
+    draw: st.DrawFn,
+    elements: Union[Sequence[El], Mapping[str, El]],
+    min_size: int = 0,
+    max_size: int = None,
+) -> st.SearchStrategy[Sequence[El]]:
+    """
+    Returns a strategy which generates sub-sequences of the input sequence.
```

Comment:
I would really like to move this (or something similar) upstream @Zac-HD !

Reply:
I'm a little dubious about the general usefulness of this strategy, and the name is pretty misleading - to me `subsequence` implies contiguity, which this *doesn't* support.  `subsets_of` might be a better name, and 

The implementation is also pretty weird - if I wanted this behaviour I'd draw `indices = st.lists(st.integers(0, len(elements)), min_size=min_size, max_size=max_size, unique=True).map(sorted)` and then use those to select the elements out.

Also, do you really need dimensions in the same order given that some can be missing from dataarrays?  e.g. if you have shapes `(a, b), (a, c)` is `(c, b)` really a problem?  If not then you can just generate `st.lists(st.sampled_from(...))` and be done.

Reply:
> Also, do you really need dimensions in the same order given that some can be missing from dataarrays? e.g. if you have shapes (a, b), (a, c) is (c, b) really a problem? If not then you can just generate st.lists(st.sampled_from(...)) and be done.

I guess it's not? Thinking it through, if I have a DataArray with coords `["x", "y", "z"]`, and I use `st.lists(st.sampled_from(...))` to create dimensions for a coordinate which returns `["y", "x"]`, then so long as I have access to the lengths along each dimension (which I do in `dim_sizes`, then I can definitely draw a `Variable` with the correct shape. So I guess I don't need `subsequences_of`!